arroyo/processing/strategies/router.py

```diff
@@ -0,0 +1,373 @@
+import logging
+from collections import deque
+from functools import partial
+from typing import Deque, Mapping, MutableMapping, Optional, Protocol, Set, Union
+
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Commit, FilteredPayload, Message, Partition, TStrategyPayload
+
+logger = logging.getLogger(__name__)
+
+
+class PartitionWatermark:
+    """
+    Keeps track of the highest committable offset when messages are processed
+    by a discrete set of independent strategies (routes here) which can
+    issue commit according to different strategies, thus out of order.
+
+    Example:
+
+    ```
+    Offsets produced:
+    Route 1:  ---- 10 --- 15 --- 20 --------------------
+    Route 2:  ------------------------- 25 --- 30 --- 35
+    ```
+
+    Route1 offsets are processed by a different strategy independently from
```

Comment:
I think we should add docstring in several functions to show what the expected/desired outcome is.